crates/red_knot_python_semantic/src/types.rs

```diff
@@ -3382,10 +3382,21 @@ impl<'db> Class<'db> {
                 explicit_metaclass_of: class_metaclass_was_from,
             }
         } else {
-            // TODO: If the metaclass is not a class, we should verify that it's a callable
-            // which accepts the same arguments as `type.__new__` (otherwise error), and return
-            // the meta-type of its return type. (And validate that is a class type?)
-            return Ok(todo_type!("metaclass not a class"));
+            let self_ty = Type::class_literal(self);
+            let bases = Type::tuple(db, self.explicit_bases(db));
+            // TODO: Should be `dict[str, Any]`
+            let namespace = KnownClass::Dict.to_instance(db);
+            // TODO: Other keyword arguments?
+
+            let outcome = metaclass.call(db, &[self_ty, bases, namespace]);
+
+            if let CallOutcome::NotCallable { not_callable_ty } = outcome {
+                return Err(MetaclassError {
+                    kind: MetaclassErrorKind::NotCallable(not_callable_ty),
+                });
+            }
+
+            return Ok(outcome.return_ty(db).unwrap().to_meta_type(db));
```

Comment:
I think we should use `return_ty_result` here for better handling of union cases?

Reply:
Hmm, our `CallOutcome` methods are in a bit of a confused state, I think tbh some refactoring is needed :/ For non-union error cases, `return_ty_result` does not emit a diagnostic (it just returns an error result), but it does for `reveal_type` and for union cases. But we don't need to refactor this in this PR.

I'm fine with directly matching on the `CallOutcome` here for now, to avoid scope creep into refactoring `CallOutcome` methods.